### PR TITLE
Foreign key constraints list

### LINKED ISSUE #943: Modify constraints API to list foreign key constraints
## Current behavior

- I can see unique and primary key constraints with a GET to `/tables/{table_id}/constraints/`, but the response does not include FK constraints.

## Desired behavior

- FK constraints should be included in the constraints response.
- Additional FK-specific details about each FK constraint should be present too, including:
    - Target columns of the FK
    - `ON DELETE` behavior
    - `ON UPDATE` behavior
    - `MATCH` behavior
    - `DEFERRED` behavior 

## Considerations
- A foreign key might be associated with more than one column, and in such cases will reference more than one column too. We're not planning to add support for creating multi-column FKs any time soon, but we do need support for displaying them, so the API needs to respond by showing all the constraint columns and all the target columns.
- Note that API consumers also need a way to associate each constraint column with its corresponding target column. If we return the constraint columns separately from the target columns, that's fine as long as the ordering of those two arrays is suitable for associating the columns by array index.
- The UI eventually want to display a specific warning in the front end when the user is deleting rows that will cascade in case of constraints with `ON DELETE CASCADE`.

## Spec Prompts
The person implementing this issue should spec out their proposed API changes and run them by the Mathesar core team. Questions the spec should answer include:
- How does the response indicate the target table and target columns?
- How does the response provide info about the FK-specific details listed above?

## Additional context
- Blocked by API changes in #898
- [Foreign Key Spec](https://wiki.mathesar.org/en/design/specs/foreign-key-constraints) -- Scenarios 2 and 3
- [Postgres docs for Constraints](https://www.postgresql.org/docs/13/ddl-constraints.html)
- [Postgres docs for `SET CONSTRAINTS`](https://www.postgresql.org/docs/13/sql-set-constraints.html)

### PR DESCRIPTION
**Pull request by silentninja**
Fixes #943 

Adds support for listing foreign key constraints to the `table/constraints/` API

**Technical details**
API schema

```typescript
interface ForeignKeyConstraint {
  id: number;
  name: string;
  type: 'foreignkey';
  referent_columns: number[];
  referent_table: number;
  onupdate: 'RESTRICT' | 'CASCADE' | 'SET NULL' | 'NO ACTION' | 'SET DEFAULT' | None;
  ondelete: 'RESTRICT' | 'CASCADE' | 'SET NULL' | 'NO ACTION' | 'SET DEFAULT' | None;
  deferrable: boolean | None;
  match: 'SIMPLE' | 'PARTIAL' | 'FULL' | None;
}
```

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by seancolsen — CHANGES_REQUESTED** [review]
@silentninja

To specify the referent column, I see that this implementation currently returns only the id.

I understand that, in theory, the id alone is sufficient to convey the association to API consumers because ids are unique across all columns of all tables. However, in practice, that's difficult for the API consumers...

Pretend I have a `moon` table with a `planet_id` column referencing a `planet` table. Issue #950 requires that, when rendering the `moon` table, the front end displays the text "planet.id" within the column header of the `planet_id` column. To render that text, I need the name of the referent table ("moon") and the name of the referent column ("id"). If I only have the column id, then I need to iterate through all tables, loading their columns one-by-one, until I find the table which contains the referent column. That's messy and slow.

I'd like for the constraints API to give me the id of the referent table too.

I'm open to various ways of solving this problem. From an API design perspective, I'm not sure what the best solution might be.

Here's one proposal, just to make this discussion as concrete as possible...

```ts
interface ForeignKeyConstraint {
  // ...
  referents: {
    table_id: number;
    column_id: number;
  }[];
  // ...
}
```

I recognize that when adding a new FK constraint via a `POST` request (to satisfy #942), the schema I'm proposing adds an unnecessary field. The API client should not need to pass `table_id` because `column_id` is sufficient. So perhaps `table_id` could be optional in the request but always present in the response.

File: mathesar/api/serializers/constraints.py [hunk 1]
```diff
@@ -1,13 +1,58 @@
...
+            'referent_columns',
+            'onupdate',
+            'ondelete',
+            'deferrable',
+            'match'
+        ]
+
+    referent_columns = serializers.PrimaryKeyRelatedField(queryset=Column.current_objects.all(), many=True)
+    onupdate = serializers.ChoiceField(choices=['CASCADE', 'DELETE', 'RESTRICT'])
+    ondelete = serializers.ChoiceField(choices=['CASCADE', 'DELETE', 'RESTRICT'])
```
Comment (mathemancer) [inline mathesar/api/serializers/constraints.py:33]: Assuming these are supposed to represent and cover the options for `ON UPDATE` and `ON DELETE` in the DB, I think these lists should be `['CASCADE', 'RESTRICT', 'NO ACTION', 'SET NULL', 'SET DEFAULT']`. Is `'DELETE'` supposed to be in place of `'SET NULL'`? See [the docs](https://www.postgresql.org/docs/13/ddl-constraints.html#DDL-CONSTRAINTS-FK).

**Review by mathemancer — CHANGES_REQUESTED** [review]
I mostly focused on Python and the DB terms. I assume you'll handle the API negotiation with @seancolsen .

File: mathesar/api/serializers/constraints.py (hunk 1 above)
Comment (silentninja) [reply mathesar/api/serializers/constraints.py]: Thanks for pointing it out @mathemancer, I was using the options listed in[alembic docs](https://alembic.sqlalchemy.org/en/latest/ops.html?highlight=foreign#alembic.operations.Operations.create_foreign_key), with the assumption that they were using generic terms to make it compatible across databases, but it looks like it isn't the case.

**Comment by silentninja** [comment]
@seancolsen I made the change to the API schema, let me know if it works for you.

**Review by mathemancer — APPROVED** [review]
Okay, I think it looks good from a DB/Python perspective.

**Review by seancolsen — APPROVED** [review]
Looks good! Thanks @silentninja

**Pull request merged**